tests/unit/test_unit_smhi.py

```diff
@@ -0,0 +1,249 @@
+        )
+        assert nearby_town[0][0] == 1
+
+    @pytest.mark.parametrize(
+        "parameter, city, distance", [(8, "Bengtsfors", None), (8, "Bengtsfors", 50)]
+    )
+    @patch("smhi.metobs.Stations.__new__")
+    @patch("smhi.metobs.Periods.__new__")
+    @patch("smhi.metobs.Data.__new__")
+    @patch("geopy.geocoders.Nominatim.__new__")
+    @patch("smhi.smhi.SMHI._find_stations_from_gps")
+    def test_find_stations_by_city(
+        self,
+        mock_find_from_gps,
+        mock_nominatim,
+        mock_data_data,
+        mock_period_data,
+        mock_station_data,
+        parameter,
+        city,
+        distance,
+    ):
+        """Unit test for SMHI _find_stations_by_city method.
+
+        Args:
+        """
+        client = SMHI()
+        _ = client._find_stations_by_city(parameter, city, distance)
+        mock_nominatim.assert_called_once()
+
+    @pytest.mark.parametrize("distance", [(0), (50)])
+    @patch("smhi.metobs.Stations.__new__")
+    @patch("smhi.metobs.Periods.__new__")
+    @patch("smhi.metobs.Data.__new__")
+    @patch("smhi.smhi.SMHI._iterate_over_time")
+    @patch("smhi.smhi.SMHI._find_missing_data")
+    @patch("smhi.smhi.SMHI._find_stations_from_gps")
+    def test_interpolate(
+        self,
+        mock_find_from_gps,
+        mock_find_missing_data,
+        mock_iterate_over_time,
+        mock_data_data,
+        mock_period_data,
+        mock_station_data,
+        distance,
+    ):
+        """Unit test for SMHI _interpolate method.
+
+        Args:
+        """
+        client = SMHI()
+        data = client._interpolate(
+            distance, mock_station_data, mock_period_data, mock_data_data
+        )
+        if distance <= 0:
+            assert data == mock_data_data
+        else:
+            mock_find_from_gps.assert_called_once()
+
+    def test_iterate_over_time(
+        self,
+    ):
+        """Unit test for SMHI _iterate_over_time method.
+
+        Args:
+        """
+        df = pd.DataFrame(
+            {
+                "date": [
+                    "2024-04-21 10:00",
+                    "2024-04-21 11:00",
+                    "2024-04-21 12:00",
+                    "2024-04-22 12:00",
+                ],
+                "Temperatur": [1, 1, 1, 12],
+            }
+        )
+        df["date"] = pd.to_datetime(df["date"])
+        df = df.set_index("date")
+
+        nearby_df = pd.DataFrame(
+            {"date": ["2024-04-22 10:00", "2024-04-22 11:00"], "Temperatur": [8, 9]}
+        )
+        nearby_df["date"] = pd.to_datetime(nearby_df["date"])
+        nearby_df = nearby_df.set_index("date")
+
+        missing_df = df[
+            df.index.to_series().diff() > df.index.to_series().diff().median()
+        ]
+
+        client = SMHI()
+
+        data = client._iterate_over_time(df, nearby_df, missing_df)
+        assert data.tail(2).iloc[0]["Temperatur"] == nearby_df.iloc[0]["Temperatur"]
+
+    def test_find_missing_data(
```

Comment:
Same here.

Reply:
Like I said, I was unable to find any way to do this with elaborate mocking of inputs to the test function, and I sort of fail to see why that would make a difference other than that it is the prettier thing to do. I figured testing the functionality of these last methods was sort of important, so I did that. The first ones are just input-output relations anyhow; whenever Metobs works they will work so I went with your original proposal to test them in a minimalistic fashion and not get stuck on writing elaborate tests for very simple methods. If you think that it is ugly, I can certainly help straighten things out, but I don't know how to do it in the way you want it (I don't even really know what it is I am supposed to do or why it matters) so you'll need to be a lot more specific and give me some actual tips beyond "google it old man", because I'm sort of stuck here. 😄

Reply:
See above two comments.